### PR TITLE
Dns feature 7012 v1

### PR DESCRIPTION
**Pull request by scrivs86**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at
   https://docs.suricata.io/en/latest/devguide/contributing/contribution-process.html
- [x] I have signed the Open Information Security Foundation contribution agreement at
   https://suricata.io/about/contribution-agreement/ (note: this is only required once)
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)
- [x] I have created a ticket at
      https://redmine.openinfosecfoundation.org/projects/suricata/issues
      (if applicable)

Link to ticket: https://redmine.openinfosecfoundation.org/issues/7012

Describe changes:
- Add dns.response sticky buffer to match on all dns response (answer) name and rdata fields (all sections)
- Add rust functions to return a reference to dns response data fields
- Unit tests to test functionality
- Updated user guide for dns sticky buffers

### Provide values to any of the below to override the defaults.

SV_BRANCH=https://github.com/OISF/suricata-verify/pull/2015

### CODE REVIEW

**Comment by scrivs86** [comment]
Note: the `detect-dns-response.c` file has many unit tests with dns packet bytes defined. I found that clang format tried to put each byte on its own line, making the code very hard to read. So I didn't format these lines with clang. I wasn't sure the best way to deal with this, but saw that some other unit tests were formatted similarly. Please advise.

**Comment by inashivb** [comment]
> Note: the `detect-dns-response.c` file has many unit tests with dns packet bytes defined. I found that clang format tried to put each byte on its own line, making the code very hard to read. So I didn't format these lines with clang. I wasn't sure the best way to deal with this, but saw that some other unit tests were formatted similarly. Please advise.

Hi! You could try to wrap those into
```
// clang-format off
....
// clang-format on
```

e.g. `src/decode.c`

File: rust/src/dns/dns.rs [hunk 1]
```diff
@@ -944,6 +944,121 @@ pub unsafe extern "C" fn SCDnsTxGetAnswerName(
...
+                return true;
+            }
+        }
+    }
+
+    false
+}
+
+#[inline]
+unsafe fn dns_get_record_rdata(data: &DNSRData, buf: *mut *const u8, len: *mut u32) -> bool {
```
Comment (jufajardini) [inline rust/src/dns/dns.rs:988]: Does the whole function have to be defined as `unsafe`? usually for our Rust side code, I see we try to restringe the `unsafe` scope as much as possible.

(nit) We usually separate the functions that are for the C API from the ones that are for the Rust side (so this could move before we start defining the C API)

**Review by jufajardini — COMMENTED** [review]
Thanks for this work!

Won't review the docs for now, as this requires some decision-making on our end, so everyone is aware that this is on our radar.

Noticed two things on the code, though.

File: rust/src/dns/dns.rs (hunk 1 above)
Comment (jasonish) [reply rust/src/dns/dns.rs:988]: Agreed.

**Comment by jasonish** [comment]
> Note: the `detect-dns-response.c` file has many unit tests with dns packet bytes defined. I found that clang format tried to put each byte on its own line, making the code very hard to read. So I didn't format these lines with clang. I wasn't sure the best way to deal with this, but saw that some other unit tests were formatted similarly. Please advise.

I also believe most of these tests can be handled by S-V test.  `dns.query` has some of these tests, but its an older keyword.  But `dns.query.name`, which is newer does not.  I wouldn't say remove them, but would also like to see the coverage that they provide also done as S-V tests, if not done already.

**Comment by jasonish** [comment]
Just some general comments, I haven't done a full-review as I do think some decisions and discussion are required.

- Naming: This keyword attempts to match on `rrname` and `rdata` fields that that resemble a name, or specific field from non-scalar rdata that contains a name like field. But should it also include stuff like `A`, and `AAAA` records? Just a point for discussion and nailing down the name.
- But then should it also apply to responses? `dns.answer.name` for example applies to requests and responses. Yes, its non-sensical to be in a request, but its still a valid DNS message.  `flow` is used to narrow down the direction, should this be similar? In which case, a name like `dns.*.name` begins to make sense, where `name` is a special name that matches on anything that looks like a hostname.
- Performance implication of all the looping done by one keyword?
- In short, the convenience of this wildcard like keyword could be useful, but we may need to think about it for a little while longer.
- Would it make sense to separate out rrname and rdata rules? `dns.*.rrname` and `dns.*.rdata.name`?  For a little more specificity, But requiring 2 rules to accomplish whats implemented here.

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 22230

File: src/detect-dns-response.c [hunk 2]
```diff
@@ -0,0 +1,1721 @@
...
+    if (buffer->initialized) {
+        return buffer;
+    }
+
+    const uint8_t *data = NULL;
+    uint32_t data_len = 0;
+
+    if (get_rdata) {
+        /* getting rdata value from resource record */
+        switch (cbdata->response_section) {
```
Comment (catenacyber) [inline src/detect-dns-response.c:90]: Could we hide this logic in rust?.. It reminds me of something in another protocol where the single index is looped over multiple vectors...

Comment (scrivs86) [reply src/detect-dns-response.c:90]: For sure. I saw in the rust code there are currently different functions for returning data from different DNS message sections, so I went with that same "style". But I could definitely have a generic rust function that takes as an argument the dns section as well as the index for the record in the section, and returns the buffer for `rdata`. The same thing would work for the `name` field, and then all these `switch/case` statements would go away.

**Comment by jasonish** [comment]
> Talking with @scrivs86 live, `dns.response` is `dns.response.domain_name` where we look for domain names in all sections : queries, answers, additional (instead of dns.query.name just looking for domain name in the queries section) 💡

Is that `rname` in DNS speak?

**Comment by scrivs86** [comment]
> > Talking with @scrivs86 live, `dns.response` is `dns.response.domain_name` where we look for domain names in all sections : queries, answers, additional (instead of dns.query.name just looking for domain name in the queries section) 💡
> 
> Is that `rname` in DNS speak?

It was great to talk to you guys in person at Suricon! Talking with you @jasonish, it seems like `dns.response.rname`  would be a decent name for this? It seems pretty clear that it is only looking at `rname` type data, but across all response sections. Let me know your thoughts, and I can work on another PR version with the feedback I have so far.

**Comment by catenacyber** [comment]
Jason will decide

**Comment by jasonish** [comment]
@scrivs86 Are you still interested in finishing this up?

- keyword: I think this should be `dns.response.rrname` meaning it looks at all known rrname types in a DNS response. `rname` is an email record, so bad comment on my part above.  More specific keywords could look like `dns.additionals...rname`, so `response` is good here I think.
- More coverage from S-V. Ideally S-V would have a test for every rrname that can be matched on

I did rebase it here https://github.com/jasonish/suricata/tree/ish/dns/feat-7012/v2, but it was pretty trivial.

**Comment by scrivs86** [comment]
> @scrivs86 Are you still interested in finishing this up?
> 
>     * keyword: I think this should be `dns.response.rrname` meaning it looks at all known rrname types in a DNS response. `rname` is an email record, so bad comment on my part above.  More specific keywords could look like `dns.additionals...rname`, so `response` is good here I think.
> 
>     * More coverage from S-V. Ideally S-V would have a test for every rrname that can be matched on
> 
> 
> I did rebase it here https://github.com/jasonish/suricata/tree/ish/dns/feat-7012/v2, but it was pretty trivial.

Hey @jasonish , yeah I would enjoy finishing this up. Thanks for the feedback! I'll change the name to what you suggest, try and address all the other feedback on this PR, as well as create some S-V tests. I'll work on a new PR version.

**Comment by jasonish** [comment]
> > @scrivs86 Are you still interested in finishing this up?
> > ```
> > * keyword: I think this should be `dns.response.rrname` meaning it looks at all known rrname types in a DNS response. `rname` is an email record, so bad comment on my part above.  More specific keywords could look like `dns.additionals...rname`, so `response` is good here I think.
> > 
> > * More coverage from S-V. Ideally S-V would have a test for every rrname that can be matched on
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I did rebase it here https://github.com/jasonish/suricata/tree/ish/dns/feat-7012/v2, but it was pretty trivial.
> 
> Hey @jasonish , yeah I would enjoy finishing this up. Thanks for the feedback! I'll change the name to what you suggest, try and address all the other feedback on this PR, as well as create some S-V tests. I'll work on a new PR version.

Great. I did some stuff after rebase and put it here: https://github.com/OISF/suricata/pull/12500 if you want to take that for any remaining items.

This has some conflicts with another aging PR I had for discrete keywords for names and rdata, and decided it would be best to build on top of this.  Will close this for now. Thanks!

**Comment by jasonish** [comment]
> > Note: the `detect-dns-response.c` file has many unit tests with dns packet bytes defined. I found that clang format tried to put each byte on its own line, making the code very hard to read. So I didn't format these lines with clang. I wasn't sure the best way to deal with this, but saw that some other unit tests were formatted similarly. Please advise.
> 
> Hi! You could try to wrap those into
> 
> ```
> // clang-format off
> ....
> // clang-format on
> ```
> 
> e.g. `src/decode.c`

Done in https://github.com/OISF/suricata/pull/12500.

File: src/detect-dns-response.c (hunk 2 above)
Comment (jasonish) [reply src/detect-dns-response.c:90]: I didn't touch this.

File: rust/src/dns/dns.rs (hunk 1 above)
Comment (jasonish) [reply rust/src/dns/dns.rs:988]: Done in https://github.com/OISF/suricata/pull/12500 as part of the rebase which made this easier.

File: src/detect-dns-response.c (hunk 2 above)
Comment (catenacyber) [reply src/detect-dns-response.c:90]: > It reminds me of something in another protocol where the single index is looped over multiple vectors...

Happens in http2 with a tx having a Vector of frames where each frame can have a vector of headers
